gateway/src/apicast/policy/3scale_batcher/README.md

```diff
@@ -0,0 +1,64 @@
+
+When the APIcast policy receives a request, it makes an 'authrep' call to
+backend. This call checks the credentials sent by APIcast, and also applies
+rate limiting over the metrics sent by APIcast. If the credentials are correct
+and rate limits not violated, backend also increases the counters of the
+metrics reported by APIcast. This counters are used both to show statistics in
+the 3scale UI and also to apply the rate limits. This means that the rate
+limiting applied will not be accurate until the counter is updated. For limits
+defined for long windows of time (hour, day, etc.) this update lag if often
+irrelevant.  However, it might be important to take it into account for limits
+defined for a small window of time (a per-minute limit, for example).
+
+This policy uses a cache for authorizations and batches reports. Also, it makes
+'authorize' and 'report' calls to backend instead of 'authrep' calls. On each
+request, the flow is as follows:
+1. The policy checks whether the credentials are cached. If they are, the
+   policy uses the cached authorization status instead of calling 3scale's
+backend. When the credentials are not cached, it calls backend and caches the
+authorization status with a configurable TTL.
+2. Instead of reporting to 3scale's backend the metrics associated with the
+   request, the policy accumulates their usages to report to backend in
+batches.
+
+Apart from that, there's a separate thread that reports to backend
+periodically. The time is configurable. This thread fetches all the batched
+reports and sends them to backend in a single call.
+
+This approach increases throughput for two reasons. First, it caches
+authorizations. This reduces the number of calls to the 3scale backend. Second,
+it batches the reports. This also helps reducing the number of calls made to
+the 3scale backend, but more importantly, it also reduces the amount of work it
+needs to do because the policy already aggregated the metrics to report.
+Suppose that we define a mapping rule that increases the metric 'hits' by one
+on each request. Suppose also that we have 1000 requests per second. If we
+define a batching period of 10 seconds, this policy will report to 3scale
+backend just a 'hits +10000' instead of 10000 separated 'hits +1'. This is very
+important, because from the 3scale backend perspective reporting a +10000 or a
++1 to its database it's the same amount of work.
+
+Of course, reporting to 3scale in batches has a trade-off. Rate limiting loses
+accuracy. The reason is that while reports are accumulated, they're not being
+sent to backend and rate limits only take into account reports that have been
+stored in the 3scale backend database. In summary, going over the defined usage
+limits is easier. The APIcast policy reports to 3scale backend every time it
+receives a request. Reports are asynchronous and that means that we can go over
+the limits for a brief window of time. On the other hand, this policy reports
```

Comment:
I'd like to explore different strategy for reporting.

Right now there is a timer to report every N seconds.

I could imagine a strategy where it would report continuously.
Lets say there is a request and it triggers backend call. Then another request comes and because the backend call is still active it would be added to shmem. Then when the backend call finishes it can collect cached reports and issue new call.

So basically it would just cache the parallel calls and make them 1:1 to backend.

All these optimizations have compromises and to chose the right one we should have some target. There is plenty ways how to do this with different tradeoffs and performance characteristics. We should define some bar we want to reach and chose the correct way to get there. 

Fully caching everything and then reporting to backend in one batch definitely has the best performance, but also the highest chance of going wrong.
This continuous reporting would not do 10x gain, maybe just 2x or 3x, but would be safer and more accurate.

Reply:
Backend report scheduling strategy is a hard problem. Definitely, trial an error with different approaches is the best strategy to come up with the best solution within our constraints.

I think it may depend on cache hit ratio. When hit ratio is high, the best approximation might be waiting as much as we can to achieve the highest batch size in allowed time window. On the other hand, when hit ratio is low, waiting does not make sense, keeping memory file descriptors and timers (through fd's) is expensive.

We could implement an algorithm that changes strategies depending on cache hit ratio that we could measure.

Reply:
On event based systems, locks are very painful with unfordable penalties in performance.

I suggest we do not use shared resources and IPC mechanisms. Each worker keeps its own state. Performance and low latency is the advantage. The drawback is increasing backend traffic since batching level is lower. Again depending on cache hit ratio. Good when cache hit ratio is high. Does not make sense on high cache miss scenarios.